### PR TITLE
Remove error_chain

### LINKED ISSUE #157: Remove error_chain with handwritten error types
error-chain is not maintained anymore and depends on Error::cause which will be deprecated in Rust 1.33.

### PR DESCRIPTION
**Pull request by dvdplm**
Implement errors manually.

Closes #157

### CODE REVIEW

File: cli/src/error.rs [hunk 1]
```diff
@@ -1,28 +1,62 @@
-#![allow(unknown_lints)]
-
-use std::io;
+use std::{fmt, io};
 use {ethabi, rustc_hex};
```
Comment (ordian) [inline cli/src/error.rs]: do we need this line?

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,11 +5,12 @@ The format is based on [Keep a Changelog].
...
 ## [Unreleased]
+### Changed
+- Removed `error_chain` and reworked errors manually (#167)
 
 ## [10.0.0] - 2020-01-08
 ### Changed
 - Replace docopt by structopt (clap) because of performance issue (#161)
 ### Fixed
 - Return an exit code 1 on failure, including wrong input parameters
 ### Dependencies
```
Comment (ordian) [inline CHANGELOG.md]:
```suggestion
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,11 +5,12 @@ The format is based on [Keep a Changelog].
...
-- Updated dependencies (https://github.com/paritytech/parity-common/pull/239)
```
Comment (ordian) [inline CHANGELOG.md]: good catch!

File: cli/src/error.rs (hunk 1 above)
Comment (dvdplm) [reply cli/src/error.rs]: Nope!

File: cli/src/main.rs [hunk 4]
```diff
@@ -176,7 +173,7 @@ fn encode_params(params: &[String], lenient: bool) -> Result<String, Error> {
 
 fn decode_call_output(path: &str, function: &str, data: &str) -> Result<String, Error> {
 	let function = load_function(path, function)?;
-	let data : Vec<u8> = data.from_hex().chain_err(|| "Expected <data> to be hex")?;
+	let data : Vec<u8> = data.from_hex()?;
```
Comment (niklasad1) [inline cli/src/main.rs]:
```suggestion
	let data: Vec<u8> = data.from_hex()?;
```

File: cli/src/main.rs [hunk 5]
```diff
@@ -196,7 +193,7 @@ fn decode_params(types: &[String], data: &str) -> Result<String, Error> {
 		.map(|s| Reader::read(s))
 		.collect::<Result<_, _>>()?;
 
-	let data  : Vec<u8> = data.from_hex().chain_err(|| "Expected <data> to be hex")?;
+	let data  : Vec<u8> = data.from_hex()?;
```
Comment (niklasad1) [inline cli/src/main.rs]:
```suggestion
	let data: Vec<u8> = data.from_hex()?;
```

File: ethabi/src/contract.rs [hunk 6]
```diff
@@ -78,21 +78,21 @@ impl Contract {
 
 	/// Creates function call builder.
 	pub fn function(&self, name: &str) -> errors::Result<&Function> {
-		self.functions.get(name).ok_or_else(|| ErrorKind::InvalidName(name.to_owned()).into())
+		self.functions.get(name).ok_or_else(|| Error::InvalidName(name.to_owned()).into())
```
Comment (niklasad1) [inline ethabi/src/contract.rs]:
```suggestion
		self.functions.get(name).ok_or_else(|| Error::InvalidName(name.to_owned()))
```

File: ethabi/src/decoder.rs [hunk 7]
```diff
@@ -28,31 +28,32 @@ fn as_u32(slice: &Word) -> Result<u32, Error> {
...
+		let res = decode_param(param, &slices, offset)?;
 		offset = res.new_offset;
 		tokens.push(res.token);
 	}
 	Ok(tokens)
 }
 
 fn peek(slices: &[Word], position: usize) -> Result<&Word, Error> {
-	slices.get(position).ok_or_else(|| ErrorKind::InvalidData.into())
+	// todo[dvdplm]: augment InvalidData to include word and pos?
```
Comment (niklasad1) [inline ethabi/src/decoder.rs]: Sounds like a good idea to me but let's create an issue and keep it out of this PR?

File: ethabi/src/event.rs [hunk 8]
```diff
@@ -130,9 +130,9 @@ impl Event {
 			0
 		} else {
 			// verify
-			let event_signature = topics.get(0).ok_or(ErrorKind::InvalidData)?;
+			let event_signature = topics.get(0).ok_or(Error::InvalidData)?;
 			if event_signature != &self.signature() {
-				return Err(ErrorKind::InvalidData.into());
+				return Err(Error::InvalidData.into());
```
Comment (niklasad1) [inline ethabi/src/event.rs]:
```suggestion
				return Err(Error::InvalidData);
```

**Pull request merged**